translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -9441,6 +9441,137 @@ Value *SPIRVToLLVM::transGLSLExtInst(SPIRVExtInst *ExtInst,
+        Args[0], ConstantFP::get(Args[0]->getType(), -1.0),
+        ConstantFP::get(Args[0]->getType(), 1.0));
+    Val =
+        getBuilder()->CreateFMul(Val, ConstantFP::get(Args[0]->getType(), 127.0));
+    Val = getBuilder()->CreateFPToSI(Val,
+                                   VectorType::get(getBuilder()->getInt8Ty(), 4));
+    return getBuilder()->CreateBitCast(Val, getBuilder()->getInt32Ty());
+  }
+
+  case GLSLstd450PackUnorm4x8: {
+    // Convert <4 x float> into unsigned normalized <4 x i8> then pack into i32.
+    Value *Val = getBuilder()->CreateFMed3(
+        Args[0], Constant::getNullValue(Args[0]->getType()),
+        ConstantFP::get(Args[0]->getType(), 1.0));
+    Val =
+        getBuilder()->CreateFMul(Val, ConstantFP::get(Args[0]->getType(), 255.0));
+    Val = getBuilder()->CreateFPToUI(Val,
+                                   VectorType::get(getBuilder()->getInt8Ty(), 4));
+    return getBuilder()->CreateBitCast(Val, getBuilder()->getInt32Ty());
+  }
+
+  case GLSLstd450PackSnorm2x16: {
+    // Convert <2 x float> into signed normalized <2 x i16> then pack into i32.
+    Value *Val = getBuilder()->CreateFMed3(
+        Args[0], ConstantFP::get(Args[0]->getType(), -1.0),
+        ConstantFP::get(Args[0]->getType(), 1.0));
+    Val = getBuilder()->CreateFMul(Val,
+                                 ConstantFP::get(Args[0]->getType(), 32767.0));
+    Val = getBuilder()->CreateFPToSI(
+        Val, VectorType::get(getBuilder()->getInt16Ty(), 2));
+    return getBuilder()->CreateBitCast(Val, getBuilder()->getInt32Ty());
+  }
+
+  case GLSLstd450PackUnorm2x16: {
+    // Convert <2 x float> into unsigned normalized <2 x i16> then pack into
+    // i32.
+    Value *Val = getBuilder()->CreateFMed3(
+        Args[0], Constant::getNullValue(Args[0]->getType()),
+        ConstantFP::get(Args[0]->getType(), 1.0));
+    Val = getBuilder()->CreateFMul(Val,
+                                 ConstantFP::get(Args[0]->getType(), 65535.0));
+    Val = getBuilder()->CreateFPToUI(
+        Val, VectorType::get(getBuilder()->getInt16Ty(), 2));
+    return getBuilder()->CreateBitCast(Val, getBuilder()->getInt32Ty());
+  }
+
+  case GLSLstd450PackHalf2x16: {
+    // Convert <2 x float> into <2 x half> then pack into i32.
+    Value *Val = getBuilder()->CreateFPTrunc(
+        Args[0], VectorType::get(getBuilder()->getHalfTy(), 2));
+    return getBuilder()->CreateBitCast(Val, getBuilder()->getInt32Ty());
+  }
+
+  case GLSLstd450PackDouble2x32:
+    // Cast <2 x i32> to double.
+    return getBuilder()->CreateBitCast(Args[0], getBuilder()->getDoubleTy());
+
+  case GLSLstd450UnpackSnorm2x16: {
+    // Unpack i32 into <2 x i16> then treat as signed normalized and convert to
+    // <2 x float>.
+    Value *Val = getBuilder()->CreateBitCast(
+        Args[0], VectorType::get(getBuilder()->getInt16Ty(), 2));
+    Val = getBuilder()->CreateSIToFP(
+        Val, VectorType::get(getBuilder()->getFloatTy(), 2));
+    Value *Multiplier =
```

Comment:
Please calculate the multiplier by CPU directly, instead of adding an additional const instruction.

Reply:
You mean the CreateFDiv? That does not generate any LLVM IR instruction if its inputs are constants. It does constant folding instead, and it uses LLVM's APFloat to calculate 1/32767, so we always get the same value whatever the host C++ compiler does.

Reply:
I think it is better to use CPU calculation if possible, avoiding LLVM does unnecessary constant folding. Write like this: Multiplier  = ConstantFP::get(Val->getType(), 1/32767.0f);

Reply:
LLPC is much slower than other competitor now, so I hope we can avoid unnecessary IR as much as possible.

Reply:
I disagree that we should use 1/32767.0f. We have no guarantee that different compilers on different OSes give the same result. Instead I could hard code the bit representation, same as I did for some constants in the trig code.

Reply:
please use bit representation as the trig code. Thanks!

Reply:
Will be resolved next time I push to rebase.